{% load variants_tags %}

{# Variant filter table header #}
<tr>
  <th class="p-0 m-0"></th> {# fold-out #}
  <th class="p-0 m-0"></th> {# rank #}
  <th class="p-0 m-0"></th> {# bookmark #}
  <th class="p-0 m-0"></th> {# databases #}
  <th></th> {# position #}
  <th></th> {# ref #}
  <th></th> {# alt #}
  <th colspan="2" class="detail-frequencies-exac">ExAC</th>
  <th colspan="2" class="detail-frequencies-gnomad-exomes">gnomAD exomes</th>
  <th colspan="2" class="detail-frequencies-gnomad-genomes">gnomAD genomes</th>
  <th colspan="2" class="detail-frequencies-thousand-genomes">1000 genomes</th>
  <th colspan="2" class="detail-frequencies-inhouse">in-house</th>
  <th class="detail-constraints-exac-pli">ExAC</th>
  <th class="detail-constraints-exac-z-mis">ExAC</th>
  <th class="detail-constraints-exac-z-syn">ExAC</th>
  <th class="detail-constraints-gnomad-pli">gnomAD</th>
  <th class="detail-constraints-gnomad-z-mis">gnomAD</th>
  <th class="detail-constraints-gnomad-z-syn">gnomAD</th>
  <th></th>{# gene #}
  {% if has_phenotype_scores %}
    <th>phenotype</th>
  {% endif %}
  {% if has_pathogenicity_scores %}
    <th>pathogenicity</th>
  {% endif %}
  {% if has_phenotype_scores and has_pathogenicity_scores %}
    <th>pheno. &amp; patho.</th>
  {% endif %}
  <th></th>{# effect #}
  {% if query_type == "case" %}
    {# pedigree members #}
    {% for member in pedigree %}
      <th></th>
    {% endfor %}
  {% else %}  {# query_type == "project" #}
    <th></th>
    <th></th>
  {% endif %}
  <th></th> {# link-outs #}
</tr>
<tr>
  <th class="p-0 m-0"></th> {# fold-out #}
  <th class="p-0 m-0"></th> {# fold-out #}
  <th class="p-0 m-0"></th> {# bookmark #}
  <th class="p-0 m-0"></th> {# databases #}
  <th>position</th>
  <th>ref</th>
  <th>alt</th>
  <th class="detail-frequencies-exac">frequency</th>
  <th class="detail-frequencies-gnomad-exomes">frequency</th>
  <th class="detail-frequencies-gnomad-genomes">frequency</th>
  <th class="detail-frequencies-thousand-genomes">frequency</th>
  <th class="detail-frequencies-exac">#hom.</th>
  <th class="detail-frequencies-gnomad-exomes">#hom.</th>
  <th class="detail-frequencies-gnomad-genomes">#hom.</th>
  <th class="detail-frequencies-thousand-genomes">#hom.</th>
  <th class="detail-frequencies-inhouse">#carriers</th>
  <th class="detail-frequencies-inhouse">#hom.</th>
  <th class="detail-constraints-exac-pli">pLI</th>
  <th class="detail-constraints-exac-z-mis">Z mis</th>
  <th class="detail-constraints-exac-z-syn">Z syn</th>
  <th class="detail-constraints-gnomad-pli">pLI</th>
  <th class="detail-constraints-gnomad-z-mis">Z mis</th>
  <th class="detail-constraints-gnomad-z-syn">Z syn</th>
  <th>gene</th>
  {% if has_phenotype_scores %}
    <th>score/rank</th>
  {% endif %}
  {% if has_pathogenicity_scores %}
    <th>score/rank</th>
  {% endif %}
  {% if has_phenotype_scores and has_pathogenicity_scores %}
    <th>score/rank</th>
  {% endif %}
  <th>effect</th>
  {% if query_type == "case" %}
    {% for member in pedigree %}
      <th style="max-width: 80px;">
        <div class="sodar-overflow-container sodar-overflow-hover">
          {{ member.patient|only_source_name }}
        </div>
      </th>
    {% endfor %}
  {% else %}  {# query_type == "project #}
    <th>sample</th>
    <th>genotype</th>
  {% endif %}
  <th></th> {# link-outs #}
</tr>
